File: src/main/java/discretemath/bool/expression/atomic/BooleanVariable.java
package discretemath.bool.expression.atomic;

import java.util.Map;
import java.util.Objects;

import discretemath.bool.expression.exception.VariableValueNotSpecifiedException;

public final class BooleanVariable implements AtomicBooleanExpression {

	private final char symbol;

	public BooleanVariable(char symbol) {
		this.symbol = symbol;
	}

	@Override
	public boolean evaluate(Map<BooleanVariable, Boolean> arguments) throws VariableValueNotSpecifiedException {
		Boolean value = arguments.get(this);

		if (value == null) {
			throw new VariableValueNotSpecifiedException(this);
		}

		return value;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		BooleanVariable that = (BooleanVariable) o;
		return symbol == that.symbol;
	}

	@Override
	public int hashCode() {
		return Objects.hash(symbol);
	}

	@Override
	public String toString() {
		return String.valueOf(symbol);
	}
}

File: src/main/scala/discretemath/bool/expression/atomic/BooleanVariable.scala
package discretemath.bool.expression.atomic;

import discretemath.bool.expression.exception.VariableValueNotSpecifiedException
import discretemath.common.VariableSymbols;

case class BooleanVariable(symbol: Char) extends AtomicBooleanExpression {

  override def getVariables: Set[BooleanVariable] = Set(this)

  @throws[VariableValueNotSpecifiedException]
  def evaluate(arguments: Map[BooleanVariable, Boolean]): Boolean = {
    return arguments.getOrElse(this, throw new VariableValueNotSpecifiedException(this))
  }

  override def isSatisfiable: Boolean = true

  override def toString: String = String.valueOf(symbol)
}

object BooleanVariable {
  val X: BooleanVariable = BooleanVariable('x')
  val Y: BooleanVariable = BooleanVariable('y')
  val Z: BooleanVariable = BooleanVariable('z')

  def getN(variablesCount: Int): IndexedSeq[BooleanVariable] = {
    return VariableSymbols.view.take(variablesCount).map(symbol => BooleanVariable(symbol)).toIndexedSeq
  }
}